File: akka-basic/src/main/java/org/darion/yaphet/java/akka/stash/ActorWithProtocol.java
package org.darion.yaphet.java.akka.stash;

import akka.actor.AbstractActorWithStash;

public class ActorWithProtocol extends AbstractActorWithStash {
    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .matchEquals("open", s -> getContext().become(receiveBuilder()
                        .matchEquals("write", ws -> System.out.println("write"))
                        .matchEquals("close", cs -> {
                            System.out.println("close");
                            unstashAll();
                            getContext().unbecome();
                        })
                        .matchAny(msg -> stash())
                        .build(), false))
                .matchAny(msg -> stash())
                .build();
    }
}

File: akka-basic/src/main/scala/org/darion/yaphet/akka/stash/ActorWithProtocol.scala
package org.darion.yaphet.akka.stash

import akka.actor.{Actor, ActorLogging, Stash}

class ActorWithProtocol extends Actor with Stash with ActorLogging {
  override def receive: Receive = {
    case "open" =>
      unstashAll()
      context.become({
        case "write" => log.info("do writing...")
        case "close" => {
          unstashAll()
          context.unbecome()
        }
      }, discardOld = false)
    case message =>
      stash()
  }
}
